core/state/snapshot/context.go

```diff
@@ -0,0 +1,223 @@
+// at the interrupted position. These iterators should be reopened from time
+// to time to avoid blocking leveldb compaction for a long time.
+func (ctx *generatorContext) openIterator(kind string, start []byte) {
+	if kind == snapAccount {
+		iter := ctx.db.NewIterator(rawdb.SnapshotAccountPrefix, start)
+		ctx.account = newSnapIter(rawdb.NewKeyLengthIterator(iter, 1+common.HashLength))
+		return
+	}
+	iter := ctx.db.NewIterator(rawdb.SnapshotStoragePrefix, start)
+	ctx.storage = newSnapIter(rawdb.NewKeyLengthIterator(iter, 1+2*common.HashLength))
+}
+
+// reopenIterators releases the held two global database iterators and
+// reopens them in the interruption position. It's aim for not blocking
+// leveldb compaction.
+func (ctx *generatorContext) reopenIterators() {
+	for i, iter := range []ethdb.Iterator{ctx.account, ctx.storage} {
+		key := iter.Key()
+		if len(key) == 0 { // nil or []byte{}
+			continue // the iterator may already be exhausted
+		}
+		kind := snapAccount
+		if i == 1 {
+			kind = snapStorage
+		}
+		iter.Release()
+		ctx.openIterator(kind, key[1:])
+	}
+}
+
+// close releases all the held resources.
+func (ctx *generatorContext) close() {
+	ctx.account.Release()
+	ctx.storage.Release()
+}
+
+// iterator returns the corresponding iterator specified by the kind.
+func (ctx *generatorContext) iterator(kind string) *snapIter {
+	if kind == snapAccount {
+		return ctx.account
+	}
+	return ctx.storage
+}
+
+// removeStorageBefore, iterates and deletes all storage snapshots starting
+// from the current iterator position until the specified account. When the
+// iterator touches the storage located in the given account range, or the
+// storage is larger than the given account range, it stops and moves back
+// the iterator a step.
+func (ctx *generatorContext) removeStorageBefore(account common.Hash) {
+	var (
+		count uint64
+		start = time.Now()
+		iter  = ctx.storage
+	)
+	for iter.Next() {
+		key := iter.Key()
+		if bytes.Compare(key[1:1+common.HashLength], account.Bytes()) >= 0 {
+			iter.Discard()
+			break
+		}
+		ctx.batch.Delete(key)
+		count += 1
+	}
+	ctx.stats.dangling += count
+	snapStorageCleanCounter.Inc(time.Since(start).Nanoseconds())
+}
+
+// removeStorageAt iterates and deletes all storage snapshots which are located
+// in the specified account range. When the iterator touches the storage which
+// is larger than the given account range, it stops and moves back the iterator
+// a step. An error will be returned if the initial position of iterator is not
+// in the given account range.
+func (ctx *generatorContext) removeStorageAt(account common.Hash) error {
+	var (
+		count int64
+		start = time.Now()
+		iter  = ctx.storage
+	)
+	for iter.Next() {
+		key := iter.Key()
+		cmp := bytes.Compare(key[1:1+common.HashLength], account.Bytes())
+		if cmp < 0 {
+			return errors.New("invalid iterator position")
+		}
+		if cmp > 0 {
+			iter.Discard()
+			break
+		}
+		ctx.batch.Delete(key)
```

Comment:
This can grow quite large if a big contract was deleted, imho we should check the batch size and flush if it gets large mid iteration. Also recreate the iterator?

Reply:
yes. we can flush the batch here. And also in order to simplify the code, we can *only* flush the batch but without persisting the generation progress marker. We usually do this in `checkAndFlush` function, but it's fine to not do it here. If panic happens, then we need to *redo* the generation since the last marker.